core/modules/gtpu_decap.cc

```diff
@@ -29,14 +32,18 @@ void GtpuDecap::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
 
   for (int i = 0; i < cnt; i++) {
     bess::Packet *p = batch->pkts()[i];
-    /* assuming that this module comes right after EthernetTrim */
-    /* pkt_len can be used as the length of IP datagram */
-    /* Trim iph->ihl<<2 + sizeof(Udp) + size of Gtpv1 header */
-    Ipv4 *iph = p->head_data<Ipv4 *>();
+    /* Trim iph->ihl<<2 + sizeof(Udp) + size of Gtpv1 header
+     */
+    Ethernet *eth = p->head_data<Ethernet *>();
+    Ipv4 *iph = (Ipv4 *)((uint8_t *)eth + sizeof(*eth));
     Gtpv1 *gtph =
         (Gtpv1 *)((uint8_t *)iph + (iph->header_length << 2) + sizeof(Udp));
+    // Don't swap lines 44 with 42, otherwise gtph->header_length()
+    // gets overwritten by ethh!!
     batch->pkts()[i]->adj((iph->header_length << 2) + sizeof(Udp) +
                           gtph->header_length());
+    memcpy((uint8_t *)gtph + gtph->header_length() - sizeof(*eth), eth,
+           sizeof(*eth));
```

Comment:
Would this simplify? and make it symmetrical with encap

```suggestion
    uint8_t *new_p = static_cast<char *> (batch->pkts()[i]->adj((iph->header_length << 2) +
                          sizeof(Udp) + gtph->header_length()));
    memcpy(new_p, eth, sizeof(*eth));
```